lib/scopes/roles/roles.go

```diff
@@ -0,0 +1,322 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package roles
+
+import (
+	"iter"
+
+	"github.com/google/uuid"
+	srpb "github.com/gravitational/teleport/api/gen/proto/go/teleport/scopedrole/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/scopes"
+	"github.com/gravitational/trace"
+)
+
+const (
+	// MaxRolesPerAssignment is the maximum number of roles@scope assignments that a given scoped role assignment
+	// resource may contain. This value is so low because our backend limits the number of keys that can be associated
+	// with a single atomic operation. Any significant increase to this value would necessitate a change to the
+	// scoped role backend model.
+	MaxRolesPerAssignment = 16
+
+	// KindScopedRole is the kind of a scoped role resource.
+	KindScopedRole = "scoped_role"
```

Comment:
any reason to define the kind here instead of in `api/types/constants.go`?

Reply:
The scopes stuff is all _very_ experimental right now, and being developed at a high cadence.  For the time being, I'm keeping everything that doesn't absolutely need to be in `api` out of `api`.  Once it stabilizes, I imagine a large amount of `lib/scopes` and `lib/scopes/roles` might end up in API.